In [4]:
import pandas as pd

In [5]:
df = pd.read_csv("../data/StudentsPerformance.csv")

# Dataset Overview
df.head(3)
df.shape
df.columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## Observations

- The dataset contains 1000 student records and 8 columns.
- There are 5 categorical features and 3 numerical features.
- The numerical columns represent scores in Mathematics, Reading, and Writing.
- No missing values were found in any column.
- The dataset appears clean and suitable for further analysis.
- Features such as gender, parental education, lunch type, and test preparation course may help identify factors affecting student performance.

In [6]:
# Data cleaning
df.isnull().sum()
df.duplicated()
df.drop_duplicates()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95
996,male,group C,high school,free/reduced,none,62,55,55
997,female,group C,high school,free/reduced,completed,59,71,65
998,female,group D,some college,standard,completed,68,78,77


## Observations

- No missing values were found in any column.
- No duplicate records were detected in the dataset.
- The dataset appears clean and does not require major preprocessing.
- The data is ready for exploratory analysis.

In [7]:
# Statistical Analysis

# Summary Statistics
df.describe()


,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


In [8]:
# Average Score of Each Subject

subject_means = df[["math score", "reading score", "writing score"]].mean()
print(subject_means)

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64


In [11]:
# Highest Average Subject
highest_avg_subject = subject_means.idxmax()
print("Highest Average Subject:", highest_avg_subject)

# Lowest Average Subject
lowest_avg_subject = subject_means.idxmin()
print("Lowest Average Subject:", lowest_avg_subject)

Highest Average Subject: reading score
Lowest Average Subject: math score


In [14]:
# Highest Score in Each Subject
print("Highest Score in Each Sub:")
print(df[["math score", "reading score", "writing score"]].max())

# Lowest Score in Each Subject
print("\nLowest Score in Each Sub:")
print(df[["math score", "reading score", "writing score"]].min())

Highest Score in Each Sub:
math score       100
reading score    100
writing score    100
dtype: int64

Lowest Score in Each Sub:
math score        0
reading score    17
writing score    10
dtype: int64


## Observations

- The average Reading score (69.17) is the highest among all subjects.
- The average Mathematics score (66.09) is the lowest among all subjects.
- Students generally perform better in Reading and Writing compared to Mathematics.
- Perfect scores (100) were achieved in all three subjects, indicating the presence of high-performing students.
- The minimum Mathematics score is 0, while the minimum Reading and Writing scores are 17 and 10 respectively, showing significant variation in student performance.

In [17]:
df["average score"] = (
    df["math score"] + 
    df["reading score"]+
    df["writing score"]
) / 3

df

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,average score
0,female,group B,bachelor's degree,standard,none,72,72,74,72.666667
1,female,group C,some college,standard,completed,69,90,88,82.333333
2,female,group B,master's degree,standard,none,90,95,93,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,49.333333
4,male,group C,some college,standard,none,76,78,75,76.333333
...,...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95,94.000000
996,male,group C,high school,free/reduced,none,62,55,55,57.333333
997,female,group C,high school,free/reduced,completed,59,71,65,65.000000
998,female,group D,some college,standard,completed,68,78,77,74.333333


In [19]:
# Performance Analysis

# Top 10 students
df["average score"].rank()

0      617.5
1      841.0
2      969.0
3       98.5
4      716.5
       ...  
995    976.5
996    233.5
997    404.0
998    662.5
999    857.0
Name: average score, Length: 1000, dtype: float64